
### Python Deep Dive: Mutability, Garbage Collection and Variable Referencing

**Variable and Memory Refenceces**

- Python calls variables as 'name'
- name points to the memory location where the value is stored (Call By object Reference as everything is an object in python)


```python
a=4    #a is the name
print(id(a)) # memory location a is pointing to
print(id(4)) # gives the same memory location when we look by value
print(hex(id(a))) # Memory location in hexadecimal
```

    94500656774464
    94500656774464
    0x55f2a613fd40


**Aliasing**


```python
a = 5
b = a
print(id(a), id(b)) # both are pointing to the same memory location
del a
print(b)
# will still show the value, as del only deletes the name 'a' which is referencing to that location,
#'b' is still referencing to the same location.
```

    94500656774496 94500656774496
    5



```python
a = 5
b = a
a = 8
print(b) # 'b' is still 5 as we have just changed where a is pointing to.
```

    5



```python
# Get how many variables are pointing to a memory location
a = "abbbabuyb"
b = a
c = b
d = a

print(a, " ", b, " ", c, " ", d) # all have same value
print(id(a), id(b), id(c), id(d)) # all are pointing to same location

# getting the count of reference
import sys
sys.getrefcount(a) #should give a result of (number of reference) + 1, as this command adds another reference to the location
# This is showing 10 here instead of 5 because of the online platform I am using to run the code, There are some other processes
# in this system which are pointing to the same location.
```

    abbbabuyb   abbbabuyb   abbbabuyb   abbbabuyb
    140536022066928 140536022066928 140536022066928 140536022066928





    10



**Garbage Collection**

This is the process of clearing memory by wiping the values which are not being referenced by any names or variables. This is automatically done periodically by a program in python called Garbage Collector. Basically if no names are referencing to a memory location and it is occupied by a value, Python will automatically clear it.

**Weird Behaviours of Python**
Python behaves weirdly when storing values in following cases:
1. getrefcount on common values
2. (-5,256) range
3. string identifiers


```python
#1. getrefcount anomaly - common values like 1,2,3 shows a big number of references as these are probably used by other processed in the system
a = 1
print(sys.getrefcount(a))
```

    3090



```python
#2. (-5,256) - If we reference to numbers in the range of (-5, 256), it references to same memory location
## But if we create something outside of this range new values get stored in a different location 
## (THIS DOES NOT HAPPEN IN THE CASE OF ALIASING, IN ALIASING REFERENCES ARE ALWAYS MADE TO THE SAME LOCATION)

a = 8769
b = a
c = b
d = c
print(id(a), id(b), id(c), id(d)) # Always Same Memory Location in case of Aliasing

a = 45
b = 45
c = 45
d = 45
print(id(a), id(b), id(c), id(d)) # Same Memory location as numbers are in the range (-5, 256)

a = 6567
b = 6567
c = 6567
d = 6567
print(id(a), id(b), id(c), id(d)) # Different memory location as numbers are outside the range (-5, 256)

# This happens because for numbers in the range (-5, 256) Python automatically stores this values as this are commonly used,
# and everytime we point to these numbers they are pointed to the same pre-created location. Python does not do that for bigger
# uncommon numbers as it will be a waste of resources
```

    140535968187920 140535968187920 140535968187920 140535968187920
    94500656775776 94500656775776 94500656775776 94500656775776
    140535968188016 140535968187568 140535968187888 140535968187952



```python
#3 Strings - For Strings if the string we are storing is a valid identifier
## then same location is referenced when we define the same string with different names
## but if the string is not a valid identifier then the value is stored in a new location
str_1 = "hamma"
str_2 = "hamma"
print(id(str_1), id(str_2)) # will be same as 'hamma' is a valid identifier
str_1 = "1not valid"
str_2 = "1not valid"
print(id(str_1), id(str_2)) # will be different as "1not valid" is not a valid identifier
```

    140535967999792 140535967999792
    140535967963888 140535967964080


**How Lists are stored**


```python
# Lists are not stored in continious memory location in python. list name points to a memory location where the pointers to the 
# memory locations of elements are stored, 
# if same element is repeated in list there is a chance that both the elements are stored only once and reference more than once

L = [1, 2, 3, 4]
print(id(L), " ", id(L[0]), " ", id(L[1]), " ", id(L[2]), " ", id(L[3])) # List elements are not stored in contiguous location
L[3] = 1 # Assigning value 1 so that there are two elements with value 1 in the list
print(L)
print(id(L[0]), " ", id(L[3])) # Both elements pointing to same location
```

    140535967961288   94500656774368   94500656774400   94500656774432   94500656774464
    [1, 2, 3, 1]
    94500656774368   94500656774368


**Mutability**

The ability to change or edit data in it's memory location. Mutability depends on the data type.

Immutable data types:
- String
- int
- float
- bool
- complex
- tuple

Mutable data types:
- List
- Dictionary
- Sets


```python
#Immutability
#Example Strings
a = "Hello"
print("Initial memory location of a: ", id(a)) # Initial memory location of a
a = a + " World"
print("Memory location after adding chars: ", id(a)) # Memory location after adding chars
# Data gets stored in a completely new location when edited and name starts pointing to that lcation
# If no other names are pointing to previous data, it gets automatically collected by garbage collector
```

    Initial memory location of a:  140535968000576
    Memory location after adding chars:  140536022013360



```python
#Mutability
# Example Lists
L = [1,2,3]
print("Memory location before editing: ", id(L)) # Memory location before editing
L.append(4) # adding 4 to the list
print("Memory location after editing: ", id(L)) # Memory location after editing remains the same
```

    Memory location before editing:  140536022104072
    Memory location after editing:  140536022104072


**Side effects of Mutation**


```python
# Original Data gets edited
# Example: List
L1 = [1,2,3]
print("List L1 before any edit: ", L1)
L2 = L1
L2.append(4)
print("list L1 after editing L2: ", L1) # 4 was not added to L1 but since lists are mutable and we have used aliasing to get the elements of list
########### any edit to L2 results in L1 being changed.
```

    List L1 before any edit:  [1, 2, 3]
    list L1 after editing L2:  [1, 2, 3, 4]


**Solution to the side effects of mutation (Cloning)**


```python
# Cloning
L1 = [1,2,3]
print("List L1 before any edit: ", L1)
L2 = L1[:] # Cloning all the elements of L1 into L2
L2.append(4)
print("list L1 after editing L2: ", L1) # L1 remains same as aliasing was not used and elements were cloned or copied to new list
```

    List L1 before any edit:  [1, 2, 3]
    list L1 after editing L2:  [1, 2, 3]



```python
# Editing mutable datatype inside an immutable datatype
a = (1,2,3,[4,5,6])
a[-1][-1] = 600 # At the end what matters is where you are editing the data
print(a)
# Reverse is also true, You can not edit an immutable type which is inside a mutable type
```

    (1, 2, 3, [4, 5, 600])



```python
# Lists on arithmatic operations changes memory location and new list gets created,
# but when we use builtin functions to edit the list the memory location does not change
a = [1,2]
b = [3,4]
c = (a,b)
print(c)
c[0][0] = 5 # Works as we are changing the list and the address of the list remains same
print(c)
# c[0] = c[0] + [6,7] will not work as the address of list a is getting changed and tuple does not allow that as its immutable
a = a + [3, 8] # a is now pointing to a new list
print(a)
print(c) # c does not change as there was no change in the memory location of c[0]
```

    ([1, 2], [3, 4])
    ([5, 2], [3, 4])
    [5, 2, 3, 8]
    ([5, 2], [3, 4])

